### PR TITLE
Fix adblock default ruleset reparsing on startup

### PR DESCRIPTION
**Pull request by atuchin-m**
Resolves https://github.com/brave/brave-browser/issues/33718
Resolves https://github.com/brave/brave-browser/issues/33532

The root of the issue is `AdBlockFiltersProviderManager` singleton. It subscribes to all the providers (in `AdBlockFiltersProviderManager::AddProvider`) => it gets notifications about an update in every provider.

Also, `AdBlockFiltersProviderManager` implements both `AdBlockFiltersProvider` and `AdBlockFiltersProvider::Observer` interfaces, acting like a decorator and proxying all the `OnChanged` to the listeners. The listeners are instances of `SourceProviderObserver` for the both engines (the manager is passed as `AdBlockFiltersProvider*` [here](https://github.com/brave/brave-core/blob/master/components/brave_shields/browser/ad_block_service.cc#L355)
As a result, both engine observers get notifications about all the updates and rebuild the combined list of every update.

The PR fixes it in safe and straightforward way: adding to `Observer::OnChange()` extra information about what kind of source was updated. This allows us to filter out unnecessary notifications.

The PR:
* implement a logic to suppress rebuilding the default engine on notifications about additional engine updates;
* adds extra traces and UMA to track the number of engine rebuilding;
* adds a browser test to verify the number of updates.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Verify that adblock engine works correctly (QA team);
2. Verify the startup performance (perf team).

### CODE REVIEW

File: components/brave_shields/browser/ad_block_service.cc [hunk 1]
```diff
@@ -96,7 +96,11 @@ AdBlockService::SourceProviderObserver::~SourceProviderObserver() {
   resource_provider_->RemoveObserver(this);
 }
 
-void AdBlockService::SourceProviderObserver::OnChanged() {
+void AdBlockService::SourceProviderObserver::OnChanged(bool is_default_engine) {
+  if (adblock_engine_->IsDefaultEngine() && !is_default_engine) {
```
Comment (atuchin-m) [inline components/brave_shields/browser/ad_block_service.cc:100]: @antonok-edm @bridiver Is it true that we have to rebuild the additional engine if the default is updated?
It doesn't clear to me, but the comments tell:
`The regional adblock engines depend on the default engine for resource loads.`
([link](https://github.com/brave/brave-core/blob/fix-adblock-ruleset-reparsing/browser/brave_shields/ad_block_service_browsertest.cc#L276-L277))

The best option to put here `adblock_engine_->IsDefaultEngine() != !is_default_engine`

File: components/brave_shields/browser/ad_block_engine.cc [hunk 2]
```diff
@@ -286,13 +307,30 @@ void AdBlockEngine::OnListSourceLoaded(const DATFileDataBuffer& filters,
 
 void AdBlockEngine::OnDATLoaded(const DATFileDataBuffer& dat_buf,
                                 const std::string& resources_json) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
```
Comment (atuchin-m) [inline components/brave_shields/browser/ad_block_engine.cc:310]: That will prevent us from ocasionally calling deserializing from UI thread.

File: browser/brave_shields/ad_block_service_browsertest.cc [hunk 3]
```diff
@@ -465,6 +466,40 @@ IN_PROC_BROWSER_TEST_F(AdBlockServiceTest,
...
+
+  // Only one new rebuild is expected. Loading the extra list must not
+  // trigger another rebuilding of the default engine.
+  histogram_tester_.ExpectTotalCount(
+      "Brave.Adblock.MakeEngineWithRules.Default", 2);
+
+  // Currently, the extra engine depends on the default, so we expect an extra
+  // unnecessary update.
+  // TODO(matuchin): remove that dependency and change 3 => 2.
+  histogram_tester_.ExpectTotalCount(
```
Comment (atuchin-m) [inline browser/brave_shields/ad_block_service_browsertest.cc:499]: To clarify the expectations in format: `{default engine updates, additional engine updates} in the begin of the test => the same of the end of the test`  
3. (before the PR) {1, 1} => {3, 3}
3. (after the PR) {1, 1} => {2, 3}
3. (what we actually want in a long term perspective) {0, 0} => {1, 1}

File: components/brave_shields/browser/ad_block_engine.cc (hunk 2 above)
Comment (bridiver) [reply components/brave_shields/browser/ad_block_engine.cc:310]: the `OnDATLoaded` method was actually removed in @antonok-edm's PR. I don't think we load dat files anymore

File: components/brave_shields/browser/ad_block_service.cc (hunk 1 above)
Comment (bridiver) [reply components/brave_shields/browser/ad_block_service.cc:100]: afaik the two engines are independent and we should not have to reload the "extras" engine when the default changes or vice versa

**Review by bridiver — APPROVED** [review]
I think this looks ok, but I'd wait on @antonok-edm to merge

File: components/brave_shields/browser/ad_block_service.cc (hunk 1 above)
Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_service.cc:100]: > `The regional adblock engines depend on the default engine for resource loads.`

That comment should be removed; that isn't the case anymore. @bridiver is correct.

Comment (antonok-edm) [reply components/brave_shields/browser/ad_block_service.cc:100]: should we do this instead?
```cpp
if (adblock_engine_->IsDefaultEngine() != is_default_engine) {
```

That way we can skip updates of the default engine for the additional engine as well.

Comment (atuchin-m) [reply components/brave_shields/browser/ad_block_service.cc]: Changed.

**Review by antonok-edm — APPROVED** [review]

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.61.45 Chromium: 118.0.5993.96 (Official Build) nightly (64-bit)
-- | --
Revision | 31c36588a37ea88c2276d901ebd20dbf2344fde3
OS | Windows 11 Version 22H2 (Build 22621.2428)
```

Basically ensured that adblocking as working when running through https://github.com/brave/brave-core/pull/20137#issuecomment-1776062895. QA will also run through some spot checks when we run through manual passes via `1.59.x`.

**Pull request merged**